### PR TITLE
persist: add some unit tests

### PR DESCRIPTION
**Pull request by danhhz**
See commits for details. The first commit isn't a test, but it was small so I snuck it in.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/persist-client/src/read.rs
Comment (danhhz) [inline src/persist-client/src/read.rs:322]: this change hangs the open_loop benchmark because it creates listeners as_of zero at startup and immediately awaits them before any writes have come in. any ideas how to proceed? everything I've come up with has been awful. https://github.com/MaterializeInc/materialize/blob/a0ec852470cfe30e5d18b41a4080ed7fcedcf1fd/src/persist-client/examples/open_loop.rs#L441-L443

File: src/persist-client/src/lib.rs
Comment (aljoscha) [inline src/persist-client/src/lib.rs:476]: Why are you using string comparison instead of something like:

```rust
            assert_eq!(
                client
                    .open::<Vec<u8>, String, u64, i64>(shard_id)
                    .await
                    .unwrap_err(),
                InvalidUsage::CodecMismatch {
                    requested: tpe("Vec<u8>", "String", "u64", "i64",),
                    actual: tpe("String", "String", "u64", "i64",),
                }
            );
```

Where `tpe()` is a helper I made up. Plus I had to add `#[cfg_attr(test, derive(PartialEq, Eq))]` on `InvalidUsage`.

The strings seem somewhat hard to maintain, but there probably is a good reason. :sweat_smile:

**Review by aljoscha — COMMENTED** [review]
Nice optimization! :raised_hands: 

Regarding the change in semantics of `listen()`: what's the reasoning behind that? It doesn't seem to be necessary for the optimization, and the test for the optimization could also be written with the old semantics. I'm asking because it seemed natural to me that `snapshot()` would block, because it does "get me the entire data up to `as_of`" while `listen()` felt more like an async stream where creating the stream at any legal `as_of` would not block but then updates would only trickle in once they are available.

File: src/persist-client/src/read.rs
Comment (aljoscha) [reply src/persist-client/src/read.rs:322]: Maybe this is an early sign that the previous behaviour of `listen()` was more ergonomic? (See my other comment)

File: src/persist-client/src/lib.rs
Comment (danhhz) [reply src/persist-client/src/lib.rs:476]: I usually lean toward matching on error message in tests because that's often how they're consumed in production. unwrap_err is a good idea, I simply forgot it existed :). I'll switch to that!

**Review by danhhz — COMMENTED** [review]
> Regarding the change in semantics of listen(): what's the reasoning behind that? It doesn't seem to be necessary for the optimization, and the test for the optimization could also be written with the old semantics. I'm asking because it seemed natural to me that snapshot() would block, because it does "get me the entire data up to as_of" while listen() felt more like an async stream where creating the stream at any legal as_of would not block but then updates would only trickle in once they are available.

I'm convinced!

File: src/persist-client/src/lib.rs
Comment (ruchirK) [inline src/persist-client/src/lib.rs:508]: hmmm this is really surprising to me because Semigroup/Monoid should be implemented for the unsigned integers and i guess it just never was a pressing need

**Review by ruchirK — APPROVED** [review]
First two commits look good to me, third commit im still reading and its taking me a bit to internalize because I'm slow this morning but don't let that block merging!

File: src/persist-client/src/read.rs
Comment (danhhz) [inline src/persist-client/src/read.rs:646]: I don't love this! some other options:
- snapshot currently clones the machine so that the methods can be `&self` instead of `&mut self` but I don't think that's super important. removing that clone would happen to make this unnecessary because the listen would inherit the state from the snapshot call
- change the `listen` call to try fetching updated state once if it's not immediately serveable
- dunno something else?

**Review by danhhz — COMMENTED** [review]
ready for another look!

File: src/persist-client/src/lib.rs
Comment (ruchirK) [reply src/persist-client/src/lib.rs:508]: opened https://github.com/TimelyDataflow/differential-dataflow/pull/368 which once it merges and we bump differential should let us test the diff param mismatch

Comment (danhhz) [reply src/persist-client/src/lib.rs:508]: oh because u64 will implement Semigroup now? nice!

**Comment by danhhz** [comment]
either of you want to take another look at this? if not, I'll resolve these conflicts (and fix my lint issue) and merge

File: src/persist-client/src/read.rs
Comment (ruchirK) [reply src/persist-client/src/read.rs:667]: looking right now! sorry about dropping this!

Comment (ruchirK) [reply src/persist-client/src/read.rs:667]: I think snapshot mutating the reader makes the most sense to me! intuitively, having a mechanism whereby a reader changes after snapshotting seems to make sense, and it seems like we're doing

```
 // Hack: Keep this method `&self` instead of `&mut self` by cloning the
 // cached copy of the state, updating it, and throwing it away
 // afterward.
 ```
 
 purely as a means to keep that method `&self`, which i think might have the rationale that its more like the expectation for the api? I don't know of any stronger reason, and given all of that, I feel like `&mut self` is a fine way forward!

File: src/persist-client/src/error.rs
Comment (aljoscha) [inline src/persist-client/src/error.rs:82]: super nit: I think timely (and Frank) already understand "beyond" as "not less than", which means "at or greater", in laymans terms.

File: src/persist-client/src/read.rs
Comment (aljoscha) [reply src/persist-client/src/read.rs:667]: When I was rewriting this test locally, to fit with the previous (and now unchanged!) semantics of listen, I changed this to  first do one `next()` call on listen, asserted against that. Then made it unreliable, and then fetched the rest of the listen events. Also slightly awkward, but doesn't require calling internal methods or changing signatures. :man_shrugging:

**Review by aljoscha — APPROVED** [review]
I had a nit and a comment. But I think this is good to merge!

File: src/persist-client/src/impl/machine.rs
Comment (aljoscha) [inline src/persist-client/src/impl/machine.rs:203]: This is super nit, but: it feels a bit weird to have these methods, because `ReadHandle` doesn't really have to call them, they're just an additional layer of verification/assertion. We could maybe put that in a comment here or maybe call this `verify_listen` or sth.

File: src/persist-client/src/read.rs
Comment (danhhz) [reply src/persist-client/src/read.rs:691]: I think it makes a lot of sense to make snapshot `&mut` and remove the machine clone (and we should consider doing that independantly), but after reverting my change to also make listen wait for as_of to be available, it seems pretty subtle for the test to rely on the fact that we call `snapshot` first. went with aljoscha's suggestion

**Review by danhhz — COMMENTED** [review]
TFTRs!!

**Pull request merged**
